File: LeetCode题目集/107. Binary Tree Level Order Traversal II(easy)-bfs法.cpp
/*
 * @Author: SourDumplings
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/binary-tree-level-order-traversal-ii/
 * @Date: 2019-03-04 17:36:39
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
  public:
	vector<vector<int>> levelOrderBottom(TreeNode *root)
	{
		vector<vector<int>> res(1000);
		queue<pair<TreeNode *, int>> q;
		q.push(make_pair(root, 0));

		if (root == nullptr)
		{
			res.clear();
			return res;
		}

		while (!q.empty())
		{
			pair<TreeNode *, int> p = q.front();
			TreeNode *node = p.first;
			int level = p.second;
			q.pop();
			res[level].push_back(node->val);
			if (node->left)
			{
				q.push(make_pair(node->left, level + 1));
			}
			if (node->right)
			{
				q.push(make_pair(node->right, level + 1));
			}
		}

		for (auto it = res.begin(); it < res.end(); it++)
		{
			if (it->empty())
			{
				res.erase(it, res.end());
			}
		}

		reverse(res.begin(), res.end());

		return res;
	}
};

File: OJ practices/LeetCode题目集/107. Binary Tree Level Order Traversal II(easy)-bfs法.java
import java.util.List;

/*
 * @Author: SourDumplings
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/binary-tree-level-order-traversal-ii/
 * @Date: 2019-03-04 18:20:04
 */

/**
* Definition for a binary tree node.
* public class TreeNode {
*     int val;
*     TreeNode left;
*     TreeNode right;
*     TreeNode(int x) { val = x; }
* }
*/
class Solution
{
    public List<List<Integer>> levelOrderBottom(TreeNode root)
    {
        List<List<Integer>> res = new LinkedList<List<Integer>>();
        Queue<TreeNode> q = new LinkedList<TreeNode>();
        if (root == null)
        {
            return res;
        }
        q.add(root);
        while (!q.isEmpty())
        {
            int levelNum = q.size();
            List<Integer> subList = new LinkedList<Integer>();
            for (int i = 0; i < levelNum; i++)
            {
                TreeNode node = q.poll();
                if (node.left != null)
                {
                    q.add(node.left);
                }
                if (node.right != null)
                {
                    q.add(node.right);
                }
                subList.add(node.val);
            }
            res.add(0, subList);
        }
        return res;

    }
}